src/internal/collection/postgres_collection.go

```diff
@@ -440,48 +561,91 @@ func (c *postgresReadOnlyCollection) Count() (int64, error) {
 	return result, c.mapSQLError(err, "")
 }
 
-func (c *postgresReadOnlyCollection) Watch(opts ...watch.Option) (watch.Watcher, error) {
-	options := watch.SumOptions(opts...)
+// This blocking function sends watch events to the client. It first sends a list of the existing elements
+// in the collection, followed by new events.
+func (c *postgresReadOnlyCollection) watchRoutine(watcher *postgresWatcher, options watch.WatchOptions, withFields map[string]string) {
+	// Do a list of the collection to get the initial state
+	lastTs := time.Time{}
+	val := cloneProtoMsg(c.template)
 
-	watcher, err := newPostgresWatcher(c.db, c.listener, c.tableWatchChannel(), c.template, nil, nil, options)
-	if err != nil {
-		return nil, err
+	lastTimestamp := func(m *model, target etcd.SortTarget) time.Time {
+		if target == SortByModRevision {
+			return m.UpdatedAt
+		}
+		return m.CreatedAt
 	}
 
-	go func() {
-		// Do a list of the collection to get the initial state
-		lastUpdated := time.Time{}
-		val := cloneProtoMsg(c.template)
-		if err := c.list(nil, &Options{Target: options.SortTarget, Order: options.SortOrder}, func(m *model) error {
-			if err := proto.Unmarshal(m.Proto, val); err != nil {
-				return errors.EnsureStack(err)
+	var bufEvent *postgresEvent
+	// Since list is not a snapshot of the DB, we break out early and hand-off
+	// event emition to the watcher if we encounter a listed record that is
+	// in the future of a buffered event
+	if err := c.list(withFields, &Options{Target: options.SortTarget, Order: etcd.SortAscend}, func(m *model) error {
+		if err := proto.Unmarshal(m.Proto, val); err != nil {
+			return errors.EnsureStack(err)
+		}
+		lastTs := lastTimestamp(m, options.SortTarget)
+
+		if bufEvent == nil {
+		loop:
+			for {
+				select {
+				case eventData := <-watcher.buf:
+					bufEvent = eventData
+				default:
+					break loop
+				}
 			}
+		}
 
-			if lastUpdated.Before(m.UpdatedAt) {
-				lastUpdated = m.UpdatedAt
+		if bufEvent != nil {
+			if bufEvent.time.Unix() <= lastTs.Unix() {
+				if err := watcher.sendInitial(bufEvent.WatchEvent(c.ctx, watcher.db, watcher.template)); err != nil {
+					watcher.sendInitial(&watch.Event{Type: watch.EventError, Err: err})
```

Comment:
I think you could just return the error here. The same logic will be applied by the error handling for the outer function.